### PR TITLE
[Merged by Bors] - dynamic scene builder

### PR DESCRIPTION
**Pull request by mockersf**
# Objective

- make it easier to build dynamic scenes

## Solution

- add a builder to create a dynamic scene from a world. it can extract an entity or an iterator of entities
- alternative to #6013, leaving the "hierarchy iteration" part to #6185 which does it better
- alternative to #6004 
- using a builder makes it easier to chain several extractions

### CODE REVIEW

File: crates/bevy_scene/src/dynamic_scene_builder.rs
Comment (alice-i-cecile) [inline crates/bevy_scene/src/dynamic_scene_builder.rs:66]: These docs should mention the behavior of extracting duplicate entities.

**Comment by alice-i-cecile** [comment]
I like this approach much better than either of the other linked PRs. I'm going to close the others in favor of this.

File: crates/bevy_scene/src/dynamic_scene_builder.rs
Comment (alice-i-cecile) [inline crates/bevy_scene/src/dynamic_scene_builder.rs:33]:
```suggestion
    /// All components registered in that world's [`AppTypeRegistry`] resource will be extracted.
```

File: crates/bevy_scene/src/dynamic_scene_builder.rs
Comment (alice-i-cecile) [inline crates/bevy_scene/src/dynamic_scene_builder.rs:52]:
```suggestion
    /// Consume the builder, producing a [`DynamicScene`].
```

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Obligatory doc nits. Once those are fixed up, this has my approval.

File: crates/bevy_scene/src/dynamic_scene_builder.rs
Comment (mockersf) [reply crates/bevy_scene/src/dynamic_scene_builder.rs:66]: Not completely convinced this should be mentioned, it would actually be very illogical to add the same entity twice in a scene. And pretty much a sure way to have a bug

Comment (alice-i-cecile) [reply crates/bevy_scene/src/dynamic_scene_builder.rs:66]: Hmm. I've often stored collections of entities in various resources. If I knew that this was idempotent WRT duplicate entities I wouldn't bother doing a uniqueness check before passing that collection into this method.

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_scene/src/dynamic_scene_builder.rs
Comment (MrGVSV) [inline crates/bevy_scene/src/dynamic_scene_builder.rs:44]: Nit: This name is just a bit wordy imo. Since this is a builder struct, I wonder if we could turn this into:
```rust
pub fn with_registry(self, type_registry: AppTypeRegistry) -> Self {
  Self { type_registry, ..self }
}
```

Although, this might allow some behavior I don't know if we'd want (maybe we do), such as swapping out the registry between extractions:

```rust
let mut builder = DynamicSceneBuilder::from_world(&world);
builder.extract_entity(entity_a);
let mut builder = builder.with_registry(&type_filter);
builder.extract_entity(entity_b);
```

File: crates/bevy_scene/src/dynamic_scene_builder.rs
Comment (MrGVSV) [reply crates/bevy_scene/src/dynamic_scene_builder.rs:66]: I think I agree with @alice-i-cecile that it should let the user know that they don't need to filter for duplicate entities before using this method. I can see why a user might think duplicates are possible, seeing as the [scene format](https://github.com/bevyengine/bevy/blob/9a597b758e5b2b2cc4355185a3fcba9564bfb104/assets/scenes/load_scene_example.scn.ron) stores entities in an array rather than a map.

File: crates/bevy_scene/src/dynamic_scene_builder.rs
Comment (MrGVSV) [inline crates/bevy_scene/src/dynamic_scene_builder.rs:112]: Nit: To reduce nesting, you could use a `continue`:
```suggestion
            if self.scene.contains_key(&entity.id()) {
                continue;
            }

            let mut entry = DynamicEntity {
                entity: entity.id(),
                components: Vec::new(),
            };

            for component_id in self.world.entity(entity).archetype().components() {
                let reflect_component = self
                    .world
                    .components()
                    .get_info(component_id)
                    .and_then(|info| type_registry.get(info.type_id().unwrap()))
                    .and_then(|registration| registration.data::<ReflectComponent>());

                if let Some(reflect_component) = reflect_component {
                    if let Some(component) = reflect_component.reflect(self.world, entity) {
                        entry.components.push(component.clone_value());
                    }
                }
            }

            self.scene.insert(entity.id(), entry);
```

File: crates/bevy_scene/src/dynamic_scene_builder.rs
Comment (MrGVSV) [inline crates/bevy_scene/src/dynamic_scene_builder.rs:115]: Is the drop necessary? Or does returning `self` without it cause a borrow error?

**Review by MrGVSV — APPROVED** [review]
Just a couple nits. Nothing blocking I don't think, though. Looks good!

File: crates/bevy_scene/src/dynamic_scene_builder.rs
Comment (mockersf) [reply crates/bevy_scene/src/dynamic_scene_builder.rs:115]: it's a borrow error without 🤷

File: crates/bevy_scene/src/dynamic_scene_builder.rs
Comment (mockersf) [reply crates/bevy_scene/src/dynamic_scene_builder.rs:44]: changing the registry between extractions could complicate things. it would mean the behaviour when extracting twice an entity with different registries is not obvious. unless this is a feature that is actually needed, I would rather avoid it

Comment (alice-i-cecile) [reply crates/bevy_scene/src/dynamic_scene_builder.rs:44]: Yeah, I agree. I think that's likely an anti-feature.

File: crates/bevy_scene/src/dynamic_scene_builder.rs
Comment (mockersf) [reply crates/bevy_scene/src/dynamic_scene_builder.rs:70]: 👍

**Comment by alice-i-cecile** [comment]
Giving this a day or two to attract reviews, then merging it unless there are other complaints.

File: crates/bevy_scene/src/dynamic_scene_builder.rs
Comment (MrGVSV) [reply crates/bevy_scene/src/dynamic_scene_builder.rs:44]: Makes sense, we can exclude this suggestion for now!

**Comment by alice-i-cecile** [comment]
The oven timer's gone off, this things is adequately baked 🥖 

bors r+

**Pull request closed without merging**
